### PR TITLE
feat: support fastify3

### PR DESCRIPTION
**Pull request by gtolarc**

### CODE REVIEW

**Comment by gtolarc** [comment]
There is a bug in graphql-upload v8.0.2. (https://github.com/jaydenseric/graphql-upload/issues/170)
Therefore, graphql-upload v8.1.0 should be used, but only supported from nodejs 8.10...

And from fastify3, req was changed to raw in the request object, and a connection object was added inside.
Therefore, it is necessary to change the logic for determining the websocket. Like this,
```
const server = new ApolloServer({
  schema,
  context: ({ raw, connection }) => {
    if (!raw) return connection.context;
  },
});
```

**Comment by gtolarc** [comment]
@abernix Please review this PR :)

**Comment by Odonno** [comment]
@gtolarc As a consumer of this project with fastify v2, can you also add a simple migration guide in a markdown file somewhere? That would great! And thank you for the migration effort BTW!

**Comment by gtolarc** [comment]
@Odonno  Added v3 migration guide documentation.

And if this PR is applied, support for node v8 or lower should be discontinued due to the compatibility problem of the fs-capacitor library used by graphql-upload. This is especially necessary to ensure compatibility with node v14.

**Comment by gtolarc** [comment]
Since [PR 3895](#3895), there is no need for a migration guide.
But, when using the fastify-fwt plugin, it should be noted that the user context is added to the request.

**Comment by dheeraj-jn** [comment]
@gtolarc, @danez Looking forward to having this PR merged.

**Comment by smolinari** [comment]
It's disappointing that even when the community comes up to the plate to bat, nobody is here to pitch the ball. 

Scott

**Comment by adamovittorio** [comment]
Maybe we can ask a review from @abernix 🙏🏻

**Comment by aecorredor** [comment]
What would it take to get this PR in? What specifically should we look for when checking the code? Willing to pitch in if anything is needed.

File: packages/apollo-server-fastify/src/ApolloServer.ts [hunk 1]
```diff
@@ -36,17 +35,14 @@ const fileUploadMiddleware = (
   uploadsConfig: FileUploadOptions,
   server: ApolloServerBase,
 ) => (
-  req: FastifyRequest<IncomingMessage>,
-  reply: FastifyReply<ServerResponse>,
+  request: FastifyRequest<any, any, any>,
```
Comment (aecorredor) [inline packages/apollo-server-fastify/src/ApolloServer.ts:38]: Should all of these be `any` or should we be able to pass a generic or anything else here?

File: packages/apollo-server-fastify/src/ApolloServer.ts [hunk 2]
```diff
@@ -168,21 +162,24 @@ export class ApolloServer extends ApolloServerBase {
             instance.addContentTypeParser(
               'multipart',
               (
-                request: IncomingMessage,
+                request: FastifyRequest<any, any, any>,
+                _payload: any,
```
Comment (aecorredor) [inline packages/apollo-server-fastify/src/ApolloServer.ts:166]: Same question abou `any` here.

File: packages/apollo-server-fastify/src/ApolloServer.ts (hunk 1 above)
Comment (gtolarc) [reply packages/apollo-server-fastify/src/ApolloServer.ts]: Yep, I modified the code in a way that does not use any type as much as possible.

**Comment by aecorredor** [comment]
@abernix could we get an official review on this one whenever you can? A lot of people want to upgrade to fastify 3 🚀

**Comment by gtolarc** [comment]
Lastly, in the process of lerna bootstrap, the twisted part was solved, and the existing dependency was kept as much as possible, and only necessary parts were remained.

The change of the file stream check in the test code is simplified because an error occurs in the node 10/12 version due to the updated graphql-upload library. @danez 

When this update is applied, node v8 compatibility should be abandoned, and I hope someone can solve this part.
But in my opinion, v8 is no longer a problem even if it is not supported.

**Comment by jsumners** [comment]
> When this update is applied, node v8 compatibility should be abandoned, and I hope someone can solve this part.
> But in my opinion, v8 is no longer a problem even if it is not supported.

Fastify v3 does not support Node 8.

**Comment by gtolarc** [comment]
https://github.com/apollographql/apollo-server/blob/main/docs/source/migration-file-uploads.md
I am talking about this. Other apollo-server implementations just skip file-upload test. But I updated the graphql-upload library to a version without problems and did not skip the test. Since this is apollo-server-core change, it goes beyond just fastify3 not supporting node8, forcing other server implementations to update to node v10 or higher.
In my opinion node v8 is now time to throw away..

**Comment by jsumners** [comment]
@gtolarc I know nothing of how this project works. I'm a maintainer of Fastify.

**Comment by gtolarc** [comment]
I personally think it's strange that apollo-server has ignored support for more than a month for successful project like fastify.

In fact, there seems to be a good alternative project already for fastify's gql support.
https://github.com/mcollina/fastify-gql

Please do not ignore the community. @abernix

**Comment by gtolarc** [comment]
If you are in a hurry, you can use [fastify-gql](https://github.com/mcollina/fastify-gql).
If you keep in mind that the cache-related minor parts and the context handling process are slightly different, quick migration is possible without any schema modification.

**Comment by jenssegers** [comment]
Not sure what's exactly changed in this PR. It mostly looks like linting changes.

**Comment by gtolarc** [comment]
graphql-upload was removed for the following reasons
https://github.com/apollographql/apollo-server/commit/3cd25e82522d85bdd5262cebf4c210cd189a16c2
final
https://github.com/apollographql/apollo-server/commit/71e89e746f8d056f46eabeb84161f1fe930a87aa
@jenssegers

**Comment by abernix** [comment]
Good news! As noted above, I cherry-picked this PR into [the `release-3.0` branch](https://github.com/apollographql/apollo-server/compare/release-3.0) last week in 71e89e746f8d056f46eabeb84161f1fe930a87aa.  Sorry for the delay in writing this up.

While 71e89e76 is the commit that will live on, for the sake of proving that the tests on this PR were all passing, I re-targeted this PR's target from the `main` branch to that branch, where I worked to align some last bits for the initial kick-off of the Apollo Server 3.0 pre-release / alpha phase over the last couple weeks.

The changes I applied in the above commit (71e89e746f8d056f46eabeb84161f1fe930a87aa) are _almost precisely_ the totality of this PR as submitted, but without some of the peripheral changes (i.e., formatting, adding `any` types).  You can see what's left on the _Files changed_ tab of this PR, but nothing substantial.

**Worth celebrating**, the above commit contribution still goes to @gtolarc and also worth applauding the efforts of those who attempted the upgrade to Fastify 2.x, notably @zuohuadong, who provided https://github.com/apollographql/apollo-server/pull/3528 and @jebh, who provided https://github.com/apollographql/apollo-server/pull/2391.  (There may also be others I'm missing, thank you!)

For what it's worth, while there are plenty of reasons why I want to start making progress on Apollo Server 3 (and we are, stay tuned!), I realize that the release of Fastify 3 was an important use case for a lot of folks, and it was one of the most within-reach wins after getting the initial AS3 release out the door, so I hope all those who stand to benefit from Fastify 3 will appreciate this.

So with that, I'll point to the availability of `apollo-server-fastify@3.0.0-alpha.3`, which was published to npm yesterday.  Please try it out!

I'll note that:
- This is an **early** alpha, but it's also early enough that most of the changes applied initially are merely to address major breaking changes (e.g., I also applied Koa upgrades).  Therefore, if it works, feel free to keep using it, but **I do recommend pinning the version** (i.e., without `^` and `~` prefixes) to avoid getting on a version that may be less stable throughout the pre-release life-time.
- This is a major breaking version, some 2.x functionality has been adjusted: Notably, GraphQL Uploads (via `graphql-upload`) and Subscriptions (`subscription-transport-ws`) are no longer integrated directly with Apollo Server, but require (not too many!) lines of code to put them in place as they were.  This aligns with the #2360 Roadmap intentions and the intentions outlined in [`ROADMAP.md`](https://github.com/apollographql/apollo-server/blob/aa9ea857404b59cab555591124e6a0091e87962f/ROADMAP.md).

[The `CHANGELOG.md`](https://github.com/apollographql/apollo-server/blob/release-3.0/CHANGELOG.md#breaking-changes) is the best place to track those changes, and there are migration guides that aim to help with the two things I just mentioned, but please feel free to help improve that documentation.  (Also, currently the instructions only touch on Express, but we could also expand them to include Fastify!  PRs to the migration guide are welcome!)

**Comment by smolinari** [comment]
Awesome! I have my system (not in production yet) using apollo-server-fastify@3.0.0-alpha.3 and it now works! Yipee!

Thank you @gtolarc for your contribution. @abernix thanks for finally giving Apollo-Server the TLC is so needs. Can't wait to see what else is now coming in the pipeline. 😁 

Scott

**Comment by ramiel** [comment]
Is it explained somewhere how to integrate graphql-upload? Thanks for the hard work!

**Comment by EdouardBougon** [comment]
@ramiel https://github.com/apollographql/apollo-server/blob/release-3.0/docs/source/migration.md#File-uploads

**Comment by ramiel** [comment]
@EdouardBougon this points to an express setup, but I'm looking for fastify. Thank you anyway

**Pull request closed without merging**